xla/service/gpu/runtime/cudnn_thunk.cc

```diff
@@ -52,6 +58,13 @@ absl::Status CuDnnThunk::Initialize(const InitializeParams& params) {
     std::string().swap(fingerprint_);
     if (result.ok()) {
       graph_->swap(*result);
+#if GOOGLE_CUDA
+      if (sdpa_dropout_seed_.has_value()) {
+        dynamic_cast<stream_executor::gpu::CudnnGraph*>(graph_->get())
+            ->InitDropoutState(params.local_device_count, *sdpa_dropout_seed_,
```

Comment:
@beckerhe Do you have an idea how we can avoid the #ifdef here? So I think in your document, one of your suggestions is to move the logic to StreamExecutor, so in this case possibly create a InitDropoutState method for DnnGraph. As far as I can tell, DnnGraph currently is only implemented by cuda backend, so to avoid that we need to cast to CudnnGraph would require specifying some default implementations for DnnGraph that return Unimplemented by default.

Reply:
2 options:

1. If `CudnnThunk` was only part of the CUDA build, then the ifdefs are just not needed. They could be removed and the target could be tagged `no_rocm` and `gpu` in the `BUILD` file to exclude it from compilation in non-GPU non-CUDA builds. Unfortunately I believe `CudnnThunk` is also needed for the ROCm build despite its name. Also this wouldn't avoid the `dynamic_cast` which I would consider a code smell.

2. The better alternative is probably to make `InitDropoutState` a pure virtual function in `DnnGraph` and then override it in `CudnnGraph` (and other implementations). This way it could always be called here no matter if it's a CUDA build or not.

Reply:
I removed the #ifdef with option 2.